src/lib/reasoners/uf.ml

```diff
@@ -1006,89 +1006,184 @@ let is_a_good_model_value (x, _) =
+  || E.equal t E.vrai || E.equal t E.faux
+  || is_optimization_op f
+  || Sy.is_internal f
+  then
+    acc
+  else
+    let xs, tys, mrepr =
+      List.fold_left
+        (fun (xs, tys, mrepr) x ->
+           let rep_x, mrepr = model_repr_of_term x env mrepr None in
+           assert (is_const_term rep_x);
+           (x, rep_x)::xs,
+           (E.type_info x)::tys,
+           mrepr
+        ) ([],[], mrepr) (List.rev xs)
+    in
+    let rep, mrepr = model_repr_of_term t env mrepr unbounded in
+    assert (is_a_good_model_value rep);
+    assert (is_const_term rep);
+    match f, xs, ty with
+    | Sy.Op Sy.Set, _, _ -> acc
+
+    | Sy.Op Sy.Get, [(_,(a,_));((_,(i,_)) as e)], _ ->
+      begin
+        match X.term_extract a with
+        | Some ta, true ->
+          let { E.f = f_ta; xs=xs_ta; _ } = E.term_view ta in
+          assert (xs_ta == []);
+          fprofs,
+          cprofs,
+          ModelMap.add (f_ta,[X.type_info i], ty) ([e], rep) carrays,
+          mrepr
+
+        | _ -> assert false
+      end
+
+    | _ ->
+      if tys == [] then
+        fprofs, ModelMap.add (f, tys, ty) (xs, rep) cprofs, carrays,
+        mrepr
+      else
+        ModelMap.add (f, tys, ty) (xs, rep) fprofs, cprofs, carrays,
+        mrepr
+
+(* A map of expressions / terms, ordered by depth first, and then by
+   Expr.compare for expressions with same depth. This structure will
+   be used to build a model, by starting with the inner/smaller terms
+   first. The values associated to the key will be their make *)
+module MED = Map.Make
+    (struct
+      type t = Expr.t
+      let compare a b =
+        let c = Expr.depth a - Expr.depth b in
+        if c <> 0 then c
+        else Expr.compare a b
+    end)
+
+let terms env = ME.fold MED.add env.make MED.empty
+
+let compute_concrete_model ?(optimized_splits=Util.MI.empty) env =
+  let bounded, pinfty, minfty =
+    Util.MI.fold
+      (fun _ord v ((bounded, pinfty, minfty) as acc) ->
+         let {Th_util.value; r; _} = v in
+         match value with
+         | Value _ ->
+           SetX.add v.Th_util.r bounded, pinfty, minfty
+         | Pinfinity  -> bounded, SetX.add r pinfty, minfty
+         | Minfinity -> bounded, pinfty, SetX.add r minfty
+         | Unknown -> acc
+      ) optimized_splits (SetX.empty, SetX.empty, SetX.empty)
+  in
+  let is_bounded = pinfty == SetX.empty && minfty == SetX.empty in
+  (* Here, we fold on each term that appears in the 'make' map,
+     starting from those with smaller depth, and we compute a concrete
+     model for it. For the objectives (if any), we check if it should
+     be +/- infinity *)
+  MED.fold
+    (fun t mk acc ->
+       if SetX.mem mk pinfty then
+         (* mk's optimum is +infinity *)
+         compute_concrete_model_of_val env t acc (Some "+oo")
+       else
+       if SetX.mem mk minfty then
+         (* mk's optimum is -infinity *)
+         compute_concrete_model_of_val env t acc (Some "-oo")
+       else
```

Comment:
We should to have a 3rd set with `pinfty` and `minfty` to collect the strict bounds and have something like `(+ n epsilon)` and `(- n epsilon)` for the strict bounds.

Reply:
Actually I have a better design for this. There is no reason to use a different ADT for the objective value during the model generation and printing objectives. I have a PR in which I use the same ADT for both. In particular, I use my `limit_kind` type to distinguish between `n`, `n + epsilon` and `n - epsilon`.  I will fix the printer in this PR also ;)